emacs_lua_complete.lua

```diff
@@ -0,0 +1,81 @@
+--    __emacs_lua_complete:
+-- This function finds and prints a limited number of matches on the
+-- lua globals tree, _G. See `lua-complete-string`. N.B.: Does not
+-- complete numeric keys (e.g. for arrays) or keys which begin with an
+-- underscore.
+-- Arguments:
+--   PARTS: a table array of dot-separated parts of the completion item
+--     (e.g. table.con would become {"table","con"}).  All but the last
+--     element must be valid subkeys in the _G heirarchy.
+--   LIBS: an array of tables of the form
+--       {var: varname, lib: libfile}
+--     for e.g. local varname=require("libfile"). This library will be
+--     require'd and searched alongside global variables.
+--   LOCALS: an array of strings parsed from local aloc,bloc = ... to
+--     complete as well
+--
+-- Example:
+-- __emacs_lua_complete({'table','ins'},{{var="xyz",lib="libfile"}},
+--                      {"frank","beans"},5000)
+--
+-- Written 2019/05 J.D. Smith
+function __emacs_lua_complete(parts,libs,locals,limit)
+   -- Setup a FIFO queue for breadth-first tree traversal
+   local queue={head=0,tail=0}
+   function queue:push(value)
+      self.tail=self.tail+1  self[self.tail]=value
+   end
+   function queue:pop()
+      self.head=self.head+1  return self[self.head]
+   end
+
+   -- Mirror the globals table
+   local globals,cnt,limit={},0,limit or 2000
+   for k,v in pairs(_G) do globals[k] = v end
+   
+   -- Add libs and locals
+   if libs then
+      for _,v in ipairs(libs) do
+	 local good,lib=pcall(require,v.lib)
+	 globals[v.var]=good and lib or {}
+      end
+   end
+   if locals then
+      for _,v in ipairs(locals) do globals[v]={} end
+   end
+
+   -- Descend the tree to the limit of dotted parts
+   local g,pre=globals,""
+   for i=1,#parts do
+      if not g then break end	-- should always exist
+      if i<#parts then		-- keep descending
+	 pre = (i > 1 and pre.."." or "")..parts[i]
+	 g = g[parts[i]]
+      else  -- final part; seed the queue with any matching tables
+	 for k,v in pairs(g) do
+	    if type(k)=="string" and k:sub(1,1) ~= "_" and k:find('^'..parts[i]) then
+	       local dpath=(i > 1 and pre.."." or "")..k
+	       print(dpath)
```

Comment:
Would it be better to refactor this function to return a list, so that it is the responsibility of the outer code to decide on how to report the list to the parent process, be it `print`, or `io.write`, or json serialization or whatever?

Reply:
Or maybe make it an iterator function somehow (I haven't written those in Lua for ages), to keep returning more and more values, and then have the wrapping function apply the `limit` and serialize the elements accordingly.

Reply:
It's a pretty small loop, often only one element, so an iterator seems a bit overkill to me.

Reply:
Ok. I don't quite like that the functions only output is a side-effect (printing to stdout) and it is impossible to tell from the outside whether or not the function has finished (you can see some lines, but don't know whether or not it is the end), but we can address it later.

Reply:
Btw, I mean the entire function that generates completions could be changed to produce an iterator, not just this loop. And then it could be consumed by another outside function that would terminate the iteration when the limit is reached. And then another one that decides how do we communicate that table to the parent process. But like I said, it is just a refactoring, we can address that later.

Reply:
Interesting idea.  That's a bit beyond my present Lua familiarity (just started with it a few weeks ago).  Hmmm... do you imagine having a special code to signal end of output?  Unless you open pipes etc. there is only one way to communicate with Emacs from asynchronous processes: printing to stdout.  The former version of writing to file and waiting an arbitrary time delay then reading it in was very brittle.  But I suppose you could write to temp file, then write the name of that file on stdout and let Emacs read the file name and load and parse the file.   Maybe a decent blended approach.